### PR TITLE
[MRG] Add a dask.distributed example

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
A simple dask.distributed example

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
For discussion and early review. Cc @TomAugspurger

File: examples/parallel/distributed_backend_simple.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import time
+import joblib
+
+def long_running_function(i):
+    time.sleep(.1)
+    return i
+
+with joblib.parallel_backend('dask.distributed',
+                             scheduler_host=client.scheduler.address):
```
Comment (TomAugspurger) [inline examples/parallel/distributed_backend_simple.py]: `client.scheduler.address` isn't always available, depending on how the type of cluster (It is available for the `LocalCluster` created automatically for `Client()`, not necessarily for others)

I believe the recommended way would be

```python
address = client.scheduler_info()['address']

with joblib.parallel_backend('dask.distributed',
                             scheduler_host=address): 
```

File: examples/parallel/distributed_backend_simple.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+"""
+Using distributed for single_machine parallel computing
+========================================================
+
+Realistic usage scenario: combining dask code with joblib code, for
+instance using dask for preprocessing data, and scikit-learn for machine
+learning.
```
Comment (TomAugspurger) [inline examples/parallel/distributed_backend_simple.py]: Would you consider "prototyping a solution, to later be run on a truly distributed cluster" a "realistic usage scenario".

That (prototyping, before moving to a cluster) and the diagnostics dashboard are my two most common use cases for using the distributed scheduler on a single-machine.

File: examples/parallel/distributed_backend_simple.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+"""
+Using distributed for single_machine parallel computing
```
Comment (glemaitre) [inline examples/parallel/distributed_backend_simple.py]: single_machine -> single machine

File: examples/parallel/distributed_backend_simple.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+client = Client()
+
+# Recover the address
+address = client.scheduler_info()['address']
+
+# This import registers the dask backend for joblib
+import distributed.joblib  # noqa
+
+###############################################################################
+# Run parallel computation using dask.distributed
```
Comment (glemaitre) [inline examples/parallel/distributed_backend_simple.py]: we could add dask to intersphinx to link to their documentation?

File: examples/parallel/distributed_backend_simple.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import time
+import joblib
+
+
+def long_running_function(i):
+    time.sleep(.1)
+    return i
+
+
+with joblib.parallel_backend('dask.distributed', scheduler_host=address):
```
Comment (glemaitre) [inline examples/parallel/distributed_backend_simple.py]: I would even add `backend='dask.distributed'` or add a small discussion after the title (it seems a bit empty there).

Comment (glemaitre) [reply examples/parallel/distributed_backend_simple.py]: But I am not 100% sure.

File: continuous_integration/circle/build_doc.sh [hunk 6]
```diff
@@ -16,7 +16,7 @@ conda update --yes --quiet conda
 conda create -n $CONDA_ENV_NAME --yes --quiet python=3
 source activate $CONDA_ENV_NAME
 
-conda install --yes --quiet pip numpy sphinx matplotlib pillow
+conda install --yes --quiet pip numpy sphinx matplotlib pillow dask distributed
```
Comment (mrocklin) [inline continuous_integration/circle/build_doc.sh]: Just `dask` suffices here.  The core package is now called `dask-core` while `dask` is a metapackage that includes `distributed` and a few other packages (like numpy, pandas, ...)

File: examples/parallel/distributed_backend_simple.py (hunk 3 above)
Comment (mrocklin) [reply examples/parallel/distributed_backend_simple.py]: FWIW we're trying to avoid referring the to code in the github.com/dask/distributed repository as `distributed`. The reason here is that it's a fairly generic term.  Instead I might recommend just using the term `dask` here, or, if preferred, `dask.distributed`.

File: examples/parallel/distributed_backend_simple.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+machine learning. In such a setting, it may be interesting to use
+distributed as a backend scheduler for both dask and joblib, to
+orchestrate well the computation.
+
+"""
+
+###############################################################################
+# Setup the distributed client
+###############################################################################
+from distributed import Client
```
Comment (mrocklin) [inline examples/parallel/distributed_backend_simple.py]: Similarly we tend to encourage `from dask.distributed import Client` in examples

File: examples/parallel/distributed_backend_simple.py (hunk 5 above)
Comment (mrocklin) [reply examples/parallel/distributed_backend_simple.py]: This should suffice if you have already created a client (requires master).  You also don't need the `address = ` line above

```python
with joblib.parallel_backend('dask'):
```

File: examples/parallel/distributed_backend_simple.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+# Typically, to execute on a remote machine, the address of the scheduler
+# would go there
+client = Client()
+
+# Recover the address
+address = client.scheduler_info()['address']
+
+# This import registers the dask backend for joblib
```
Comment (aabadie) [inline examples/parallel/distributed_backend_simple.py]: import**s** and registers

Comment (aabadie) [reply examples/parallel/distributed_backend_simple.py]: s/dask/dask.distributed/

File: examples/parallel/distributed_backend_simple.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+def long_running_function(i):
+    time.sleep(.1)
+    return i
+
+
+with joblib.parallel_backend('dask.distributed', scheduler_host=address):
+    joblib.Parallel(n_jobs=2, verbose=100)(
+        joblib.delayed(long_running_function)(i)
+        for i in range(10))
+    # We can check that joblib is indeed using the dask.distributed
```
Comment (aabadie) [inline examples/parallel/distributed_backend_simple.py]: reword without `we`

File: examples/parallel/distributed_backend_simple.py (hunk 3 above)
Comment (GaelVaroquaux) [reply examples/parallel/distributed_backend_simple.py]: I'll use dask.distributed.

This naming thing is quite confusing (I can't blame you, it's an error that I been into over and over, first with enthought.mayavi => mayavi, and later with scikits.learn => scikit-learn (imported as sklearn)). However, it will confuse the users, and even myself, as it makes the difference and the boundary between projects quite blurry.

File: examples/parallel/distributed_backend_simple.py (hunk 4 above)
Comment (GaelVaroquaux) [reply examples/parallel/distributed_backend_simple.py]: Yey! Great idea!

File: examples/parallel/distributed_backend_simple.py (hunk 5 above)
Comment (GaelVaroquaux) [reply examples/parallel/distributed_backend_simple.py]: Now backend='dask' is very confusing, because, as @TomAugspurger was explaining to me, by default dask doesn't use the distributed backend, but a threading one.

I'll wait for a release of distributed to update this example, as I would like it to run with released version.

File: examples/parallel/distributed_backend_simple.py (hunk 7 above)
Comment (GaelVaroquaux) [reply examples/parallel/distributed_backend_simple.py]: This is terribly confusing, you realize. I haven't looked at the codebases of the package, but my mental model right now is a bit lost as to what package does what.

File: examples/parallel/distributed_backend_simple.py (hunk 8 above)
Comment (GaelVaroquaux) [reply examples/parallel/distributed_backend_simple.py]: I think that that was correct: import is used as a noun here.

File: examples/parallel/distributed_backend_simple.py (hunk 9 above)
Comment (glemaitre) [reply examples/parallel/distributed_backend_simple.py]: remove `we can` :)

**Comment by GaelVaroquaux** [comment]
I can't get the intersphinx mapping to work. Is distributed.Client documented in what is captured by intersphinx?

File: examples/parallel/distributed_backend_simple.py (hunk 1 above)
Comment (mrocklin) [reply examples/parallel/distributed_backend_simple.py]: client.scheduler.address *is* always available, in this case `scheduler` is actually the connection object, not the scheduler itself.  In either event though you don't need either.

Also, if relevant, I hope to release dask.distributed within a week.

File: examples/parallel/distributed_backend_simple.py (hunk 4 above)
Comment (mrocklin) [reply examples/parallel/distributed_backend_simple.py]: I would love to see this happen.  I have no experience with intersphinx myself but have seen it in use more often more recently and have liked what I've seen.

**Comment by GaelVaroquaux** [comment]
> Also, if relevant, I hope to release dask.distributed within a week.

Cool!

I think that I'd like to merge this example first, and modify it as soon
as you release dask.distributed.

I also hope that we'll release an alpha of the new joblib soon after.
That way, we'll all be in production soon.

File: examples/parallel/distributed_backend_simple.py (hunk 5 above)
Comment (mrocklin) [reply examples/parallel/distributed_backend_simple.py]: Right, but you'll never use this with the legacy threaded scheduler, it isn't sufficiently flexible to handle Joblib's dynamism.  There is only one relevant use case for Dask here, and it's the newer scheduler.

File: examples/parallel/distributed_backend_simple.py (hunk 7 above)
Comment (mrocklin) [reply examples/parallel/distributed_backend_simple.py]: It's not that confusing if you're coming to it fresh (which most people are).  It's really only a pain for the old hands who were around when we called the thing `distributed` on its own.

**Comment by GaelVaroquaux** [comment]
> Right, but you'll never use this with the legacy threaded scheduler, it isn't
> sufficiently flexible to handle Joblib's dynamism. There is only one relevant
> use case for Dask here, and it's the newer scheduler.

OK, and the threaded scheduler is going away? That would explain part of
my confusion.

**Comment by mrocklin** [comment]
> OK, and the threaded scheduler is going away? That would explain part of
my confusion.

No it's staying around.  It's useful if you don't have tornado, are allergic to dependencies (it's stdlib only), or if you're doing relatively straightforward dask.array work.  The newer scheduler is generally a more robustly a good choice though.

**Comment by GaelVaroquaux** [comment]
I found the problem with intersphinx: Client is documented as
distributed.client.Client:
https://github.com/dask/distributed/blob/master/docs/source/api.rst

while we use it as distributed.Client.

I believe that the "currentmodule" should be changed in the file above.

**Comment by mrocklin** [comment]
Is there a way for us to flatten the namespace on our end?

**Comment by GaelVaroquaux** [comment]
> are allergic to dependencies (it's stdlib only),

Thanks for having my health in mind :).

**Comment by mrocklin** [comment]
Ha!

**Comment by GaelVaroquaux** [comment]
> Is there a way for us to flatten the namespace on our end?

In the API documentation file, just use as a "currentmodule" "dask": the
currentmodule will define what sphinx considers is the public path of the
objects.

File: examples/parallel/distributed_backend_simple.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import joblib
+
+
+def long_running_function(i):
+    time.sleep(.1)
+    return i
+
+
+with joblib.parallel_backend('dask.distributed', scheduler_host=address):
+    joblib.Parallel(n_jobs=2, verbose=100)(
```
Comment (ogrisel) [inline examples/parallel/distributed_backend_simple.py]: In current master, there is no need to put `n_jobs=2` here anymore.

File: examples/parallel/distributed_backend_simple.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        joblib.delayed(long_running_function)(i)
+        for i in range(10))
+    # Check that joblib is indeed using the dask.distributed
+    # backend
+    print(joblib.Parallel(n_jobs=1)._backend)
```
Comment (ogrisel) [inline examples/parallel/distributed_backend_simple.py]: This is no longer necessary, I have added the active backend name in the verbose output of the call to `Parallel`.

**Comment by GaelVaroquaux** [comment]
I've addressed all issues, and CI is green.

Can I have merge?

File: examples/parallel/distributed_backend_simple.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+def long_running_function(i):
+    time.sleep(.1)
+    return i
+
+
+###############################################################################
+# The verbose messages below show that the backend is indeed the
+# dask.distributed one
+with joblib.parallel_backend('dask.distributed', scheduler_host=address):
```
Comment (mrocklin) [inline examples/parallel/distributed_backend_simple.py]: Adding `scheduler_host=address` is no longer strictly necessary.  Dask will use the most recently created Client by default.

Leaving it in is ok too.

**Comment by GaelVaroquaux** [comment]
Merging this guy.
